canvas.py

```diff
@@ -184,99 +292,284 @@ def _download_from_folders(self, course_id, course_name) -> bool:
     def _download_from_modules(self, course_id, course_name) -> bool:
         modules_list = self.get_modules(course_id)
 
-        for module in modules_list:
+        if not modules_list or "errors" in modules_list:
+            print_c(f"Warning: Modules are not available for course {course_name}. Skipping.", "error", 1)
+            return False
 
-            if not module["items_count"]:
+        for module in modules_list:
+            if not module.get("items_count"):
                 continue
 
             module_items = self.get_modules_items(course_id, module["id"])
-
             if "errors" in module_items:
-                return False
+                print_c(f"Warning: Unable to retrieve module items for {module['name']} in course {course_name}.", "error", 2)
+                continue
 
-            # TODO: A module can have a name that is not a valid path
-            module_path = [course_name, module["name"].strip().replace("/", "&")]
+            module_path = [course_name, module["name"].strip().replace("/", "&").replace(":", "-")]
             print_c("[M] " + module["name"], "item", 1)
 
             for item in module_items:
+                if not isinstance(item, dict) or "type" not in item:
+                    print_c("Error in _download_from_modules: invalid item, skipping", "error", 2)
+                    continue  # Skip invalid objects
+
                 if item["type"] == "File":
                     file_obj = self.get_file_from_id(course_id, item["content_id"])
                     folder_obj = self.get_folder_from_id(course_id, file_obj["folder_id"])
-                    if "full_name" in folder_obj:
-                        current_folder_path = [course_name] + folder_obj["full_name"].split("/")[1:]
-                    else:
-                        current_folder_path = module_path
-                    self._download_file(
-                        file_obj["url"], current_folder_path, file_obj["display_name"]
-                    )
+                    current_folder_path = ([course_name] + folder_obj["full_name"].split("/")[1:]) if "full_name" in folder_obj else module_path
+                    if not isinstance(file_obj, dict) or "url" not in file_obj:
+                        print_c("Error in module item file downloader: invalid item, skipping", "error", 2)
+                        continue  # Skip invalid objects
+                    self._download_file(file_obj["url"], current_folder_path, file_obj["display_name"])
+
                 elif item["type"] == "ExternalUrl":
-                    download_url = get_external_download_url(item["external_url"])
-                    if download_url:
-                        self._download_file(download_url, module_path)
+                    if self._is_canvas_url(item["external_url"]):
+                        self._download_canvas_page(course_id, item["external_url"], course_name)
```

Comment:
Here there was a method to download files from Google Drive, since at UC Chile it was common for courses to link to files there. I would keep the `get_external_download_url` check, or at least add a comment to say that here is where one could add a custom downloader for external services.